File: c++/300_LongestIncreasingSubsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
    	// return func1(nums);
    	return func2(nums);
    }

    // ** dp[i] represents the LIS end at i
    int func1(vector<int>& nums) {
    	int N = nums.size();
    	if (N <= 0) return 0;

    	vector<int> dp(N, 1);

  		int res = 1;
    	for (int i = 1; i < N; i++) {
    		for (int j = i-1; j >= 0; j--) {
    			if (nums[i] > nums[j]) {
    				dp[i] = max(dp[i], 1 + dp[j]);
    			}
    		}
    		res = max(res, dp[i]);
    	}

    	return res;
    }

    // ** Reference, using a array called 'tails' to store
    // ** the smallest tail with the len of LIS which
    // ** may 1, 2, ..., N-1
    int func2(vector<int>& nums) {
    	int N = nums.size();
    	vector<int> tails(N+1, -1);

    	bool doPrint = false;
    	
    	int len = 1;
    	for (auto num : nums) {

            int lo = 1, hi = len;

    		if (doPrint) {
    			printf("beg: lo=%d, hi=%d\n", lo, hi);
    		}

    		while (lo < hi) {
    			int mid = (lo + hi) / 2;
    			if (tails[mid] < num) {
    				lo = mid + 1; 
    			} else {
    				hi = mid;
    			}
    		}
    		tails[lo] = num;

    		if (doPrint) {
    			printf("end: lo=%d, hi=%d\n", lo, hi);
    		}

    		if (lo == len) {
    			len++;
    		}

    		if (doPrint) {
    			for (int i = 0; i <= N; i++) {
	    			printf("length = %d with tail = %d\n", i, tails[i]);
	    		}
	    		printf("==========\n");
    		}
    		
    	}


    	return len - 1;
    }
};

File: golang/300_LongestIncreasingSubsequence.go
func lengthOfLIS(nums []int) int {
	// return func1(nums);
	// return func2(nums);
	return func3(nums);
}

//** sort + LCS
func func1(nums []int) int {
	if len(nums) <= 0 {
		return 0;
	}

	sorted := make([]int, len(nums));
	for i := 0; i < len(nums); i++ {
		sorted[i] = nums[i];
	}
	sort.Ints(sorted);

	unique := make([]int, 1, len(nums));
	unique[0] = sorted[0];
	for i := 1; i < len(sorted); i++ {
		if sorted[i] != unique[len(unique)-1] {
			unique = append(unique, sorted[i]);
		}
	}

	return lengthOfLCS(unique, nums);
}
func lengthOfLCS(nums1 []int, nums2 []int) int {
	m := len(nums1);
	n := len(nums2);

	ret := 0;
	memo := make([][]int, m+1);
	for i := 0; i < m+1; i++ {
		memo[i] = make([]int, n+1);
	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if nums1[i] == nums2[j] {
				memo[i+1][j+1] = memo[i][j] + 1;
			} else {
				memo[i+1][j+1] = max(memo[i][j+1], memo[i+1][j]);
			}
			ret = max(ret, memo[i+1][j+1])
		}
	}

	return ret;
}
func max(a int, b int) int {
	if a > b {
		return a;
	} else {
		return b;
	}
}

//** standard dp
func func2(nums []int) int {
	n := len(nums);
	memo := make([]int, n);
	for i := 0; i < n; i++ {
		memo[i] = 1;
	}
	ret := 0;

	for i := n-1; i >= 0; i-- {
		for j := i+1; j < n; j++ {
			if nums[i] < nums[j] {
				memo[i] = max(memo[i], memo[j] + 1);
			}
		}
		ret = max(ret, memo[i]);
	}

	return ret;
}

//** O(nlogn) binary-search solution, interesting
func func3(nums []int) int {
	n := len(nums);
	arr := make([]int, 0);

	for i := n-1; i >= 0; i-- {
		idx := findLargestLessOrEqualThanTarget(arr, nums[i]);
		if idx >= len(arr) {
			arr = append(arr, nums[i])
		} else {
			arr[idx] = nums[i];
		}
	}

	return len(arr);
}

/* 
 * nums is inverted order and return the index that nums[index] is the
 * largest element <= target
 * eg: [19,18,17,16,10], find target | nums[lo] | nums[hi] | ret:
 * 20 | nums[0]=19 | nums[-1]=  | 0
 * 19 | nums[0]=19 | nums[0]=19 | 0
 * 12 | nums[4]=10 | nums[3]=16 | 4
 * 09 | nums[5]=   | nums[4]=10 | 5
 */
func findLargestLessOrEqualThanTarget(nums []int, target int) int {
	lo := 0;
	hi := len(nums) - 1;

	for (lo <= hi) {
		mid := (lo + hi) / 2;
		if nums[mid] == target {
			return mid;
		} else if nums[mid] > target {
			lo = mid + 1;
		} else {
			hi = mid - 1;
		}
	}

	return lo;
}
